aten/src/ATen/native/cuda/ScatterGatherKernel.cu

```diff
@@ -154,7 +154,85 @@ struct _cuda_scatter_gather_internal_kernel {
 
     _launch_scatter_gather_kernel<num_threads(), thread_work_size()>(iter.numel(), loop);
   }
-}; // struct _cuda_scatter_fill_internal_kernel
+}; // struct _cuda_scatter_gather_internal_kernel
+
+// When scatter is called and the index tensor
+// has any dimensions larger than the source tensor
+// we use this subroutine instead of the default one.
+template <typename scalar_t>
+struct _cuda_scatter_large_index_internal_kernel {
+  template <typename func_t>
+  void operator() (
+    TensorIterator& iter,
+    int64_t index_size,
+    int64_t index_stride,
+    const IntArrayRef index_shape,
+    const IntArrayRef index_strides,
+    const IntArrayRef src_shape,
+    const IntArrayRef src_strides,
+    int64_t numel,  // Do not use `const` qualifier here as it may cause issue in cuda 11.6.x. See #75434, #75545
+    const func_t& f
+  ) {
+    if (!iter.can_use_32bit_indexing()) {
+      for (auto& sub_iter : iter.with_32bit_indexing()) {
+        _cuda_scatter_large_index_internal_kernel<scalar_t>()(
+          sub_iter, index_size, index_stride, index_shape, index_strides, src_shape, src_strides, numel, f
+        );
+      }
+      return;
+    }
+
+    char* self_ptr = (char*)iter.data_ptr(0);
+    char* src_ptr = (char*)iter.data_ptr(1);
+    char* index_ptr = (char*)iter.data_ptr(2);
+
+    // We send the relevant index and source metadata
+    // to the device
+    int ndim = index_shape.size();
+    int64_t *index_shape_device, *index_strides_device, *src_shape_device, *src_strides_device;
```

Comment:
That's exactly the sort of thing I expected I'd miss/not know about, thank you! I'll look into changing that.

Reply:
Okay, so, looking into it, I am not quite sure which of the various caching allocators I should use. There seem to be several options, the most likely-looking one being the `CUDACachingAllocator` one, but looking through examples in the codebase to check for usage does not make it immediately obvious how to then copy the data from host to device. I don't suppose you'd know of some resource or guide or particularly enlightening snippet that could point at what the best practices are?

Reply:
Why not allocating a single buffer with a Tensor? You can get it's head with `t.data_ptr<type>()`.

Reply:
I'm not sure what you mean by that. What does seem necessary to me is a way to have an ordered list of strides and sizes for index and source, given that there's (as far as I could tell) no way to recover the information of what index a given offset is equivalent to in the original tensor with just the information otherwise provided to the loop. But there might be a way to access those values from the iterator itself or something in there that I missed, in which case passing it to device memory the way I did would be entirely unnecessary.